### PR TITLE
API function to replace workspace being shown by an instrument window

### LINKED ISSUE #29846: Make it possible to replace the workspace being shown by the instrument view
### Expected behavior
As a first step to develop an instrument view based widget for the ILL, a way to replace the workspace being shown by the instrument window without having to close it is needed.

This will take the instrument window object and the new workspace as arguments, and upon call replace the view and reset all the parameters in the window.

### PR DESCRIPTION
**Pull request by MathieuTillet**
**Description of work.**
This PR adds an API function to replace the workspace that is being shown by an instrument window.
When called, the view is reset, the color and integration range are updated and the window's title is changed to new values. Only the instrument render mode is kept the same.
This is a step in the development of a future widget, and as such is not really useful on its own.

**To test:**
The instrument widget not being compliant with the MVP model, this function unfortunately needs manual testing.
The following strange syntax regarding threads is due to issues with the way they communicate in Workbench's embedded python interpreter.

To test: 
- Add `ExternalData/Testing/Data/UnitTest/ILL/D33/` to the list of user directories.
- Run the following script. During the pauses provided by the sleeps, manipulate the instrument view and check that everything - integration range, color bar, view direction, pick tab - is reset when the view changes.
```
from mantid.simpleapi import CreateSampleWorkspace
from mantidqt.widgets.instrumentview.presenter import InstrumentViewPresenter
from mantidqt.utils.qt.qappthreadcall import QAppThreadCall

from time import sleep

def show_and_replace_instrument_view(first_ws, first_name, second_name, new_window_name=None):
    # For thread related reasons, the instrument view and the replace_ws calls need to be made on the same function,
    # using QAppthreadCall.
    # This limitation is only present for scripts run in the workbench's embedded script window, and should not be of concern
    # in the final use case of this function.
    
    iv = QAppThreadCall(InstrumentViewPresenter)(first_ws)
    QAppThreadCall(iv.show_view)()
    
    while True:
        sleep(8)
        QAppThreadCall(iv.replace_workspace)(second_name, new_window_name)
        sleep(2)
        QAppThreadCall(iv.replace_workspace)(first_name)
    
first = Load(Filename="093410", OutputWorkspace="093410")
second = LoadEmptyInstrument(InstrumentName="d16", OutputWorkspace="ws")

show_and_replace_instrument_view(first, "093410", "ws", "Empty_D16")
```

Fixes #29846 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/instrumentview/inc/MantidQtWidgets/InstrumentView/InstrumentWidget.h [hunk 1]
```diff
@@ -43,6 +43,8 @@ class QDragEnterEvent;
 class QDropEvent;
 class QStackedLayout;
 class QSettings;
+class QVBoxLayout;
+class QHBoxLayout;
```
Comment (DannyHindson) [inline qt/widgets/instrumentview/inc/MantidQtWidgets/InstrumentView/InstrumentWidget.h:47]: Not sure you need this forward declaration of QHBoxLayout - but not worth worrying about if this is the only change

**Review by DannyHindson — COMMENTED** [review]
I've tested this out. I found two things relating to the Render tab when running the test script with the sleep commands in it:

1) if I change the drop down near the bottom left of the screen from Linear to SymmetricLog10 when the instrument is 093410, then this change sticks whenever the view switches back to 093410. Is that correct\expected? The testing instructions made it sound like all the settings should be reset
2) If I change the drop down near the top left of the tab from "Full 3D" to "Cylindrical X" while instrument 093410 is displayed then workbench crashes with a segmentation fault a few seconds later (probably when the next replace workspace command is run).

I've also added one v pedantic code review comment about a forward declaration but not essential to do this if that's the only change

**Comment by MathieuTillet** [comment]
These changes take care of the problems you pointed, but a recent change in master adds a bug that makes the replace function unusable ( https://github.com/mantidproject/mantid/issues/30005 ), so perhaps it would be better to wait for it to be fixed.

**Comment by DannyHindson** [comment]
> These changes take care of the problems you pointed, but a recent change in master adds a bug that makes the replace function unusable ( #30005 ), so perhaps it would be better to wait for it to be fixed.

@MathieuTillet - I haven't seen a fix go in for issue #30005 so I haven't retested this item yet. Let me know though if I've missed the fix

**Comment by MathieuTillet** [comment]
@DannyHindson The bug has just been fixed, so you can test again.

**Comment by DannyHindson** [comment]
I've retested this and the points I raised are all fixed now thanks. Just one question - the choice of view on the Render tab (Full 3D, Cylindrical X, Cylindrical Y etc) sticks when the instrument view switches after each sleep in the test script. Is that correct? In the testing instructions you'd mentioned that:
everything - integration range, color bar, view direction, pick tab - is reset when the view changes.

**Comment by MathieuTillet** [comment]
Fair remark, it now resets to the default view of the instrument, which is more logical

**Comment by DannyHindson** [comment]
Thanks. I see the PR has hit a strange error from the ClangFormat job on Jenkins that doesn't look to be related to your changes. Looks like another PR got merged into master with some format problems in it (https://github.com/mantidproject/mantid/pull/29967). Not v obvious to me what went wrong though

**Review by DannyHindson — APPROVED** [review]
Thanks for making all the changes. Have just retested and all looks good

**Comment by gvardany** [comment]
@MathieuTillet  is there a flag for preserving our resetting the projection/viewpoint? I am happy to merge this, and leave that to a subsequent PR. Concerning your instructions to test, since now there is [this](https://github.com/mantidproject/mantid/blob/master/qt/python/mantidqt/widgets/instrumentview/api.py#L25), do I get right that this will now work:
```
iv = get_instrumentview(ws1)
iv.replace_workspace(ws2)
```

**Comment by MathieuTillet** [comment]
The flag is not yet done, no.
And yes, this now works (and is far more readable)

**Pull request merged**